services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/schema/AnnotationsTranslatorImpl.java

```diff
@@ -0,0 +1,379 @@
+import org.sagebionetworks.schema.adapter.org.json.JSONObjectAdapterImpl;
+import org.sagebionetworks.schema.adapter.org.json.JsonDateUtils;
+import org.sagebionetworks.util.ValidateArgument;
+import org.springframework.stereotype.Service;
+
+import com.google.common.collect.Lists;
+
+@Service
+public class AnnotationsTranslatorImpl implements AnnotationsTranslator {
+
+	private static final String ID = "id";
+	private static final String ETAG = "etag";
+	public static final String CONCRETE_TYPE = "concreteType";
+
+	@Override
+	public JSONObject writeToJsonObject(Entity entity, Annotations annotations) {
+		ValidateArgument.required(entity, "entity");
+		ValidateArgument.required(annotations, "annotations");
+		JSONObject jsonObject = new JSONObject();
+		writeAnnotationsToJSONObject(annotations, jsonObject);
+		JSONObjectAdapterImpl adapter = new JSONObjectAdapterImpl(jsonObject);
+		try {
+			// write the entity second to override any conflicts.
+			entity.writeToJSONObject(adapter);
+		} catch (JSONObjectAdapterException e) {
+			throw new IllegalStateException(e);
+		}
+		return jsonObject;
+	}
+
+	@Override
+	public Annotations readFromJsonObject(Class<? extends Entity> entityClass, JSONObject jsonObject) {
+		ValidateArgument.required(entityClass, "entity");
+		ValidateArgument.required(jsonObject, "jsonObject");
+		validateHasKeys(Lists.newArrayList(ID, ETAG, CONCRETE_TYPE), jsonObject);
+		String entityId = jsonObject.getString(ID);
+		if (!entityClass.getName().equals(jsonObject.get(CONCRETE_TYPE))) {
+			throw new IllegalArgumentException(
+					"The value of 'concreteType' does not match the type of Entity: '" + entityId + "'");
+		}
+		Annotations annotations = new Annotations();
+		annotations.setId(entityId);
+		annotations.setEtag(jsonObject.getString(ETAG));
+		Map<String, AnnotationsValue> map = new LinkedHashMap<String, AnnotationsValue>();
+		annotations.setAnnotations(map);
+		jsonObject.keySet().stream().filter(key -> canUseKey(entityClass, key)).forEach((key) -> {
+			AnnotationsValue annValue = getAnnotationValueFromJsonObject(key, jsonObject);
+			map.put(key, annValue);
+		});
+		return annotations;
+	}
+
+	void validateHasKeys(List<String> expectedKeys, JSONObject jsonObject) {
+		for (String expectedKey : expectedKeys) {
+			if (!jsonObject.has(expectedKey)) {
+				throw new IllegalArgumentException("Expected JSON to include key: '" + expectedKey + "'");
+			}
+		}
+	}
+
+	/**
+	 * Read a single AnnotationValue for the given key from the given JSONObject.
+	 * 
+	 * @param key
+	 * @param jsonObject
+	 * @return
+	 */
+	AnnotationsValue getAnnotationValueFromJsonObject(String key, JSONObject jsonObject) {
+		return Stream.of(attemptToReadAsJSONArray(key, jsonObject), attemptToReadAsDouble(key, jsonObject),
+				attemptToReadAsTimestamp(key, jsonObject), attemptToReadAsLong(key, jsonObject),
+				attemptToReadAsString(key, jsonObject)).filter(Optional::isPresent).findFirst().get().get();
```

Comment:
The findFirst() might return an empty optional, we should consider doing an orElseThrow().